### DISCUSSION TITLE
How to assign an expression of a specific type to a list of a union

### DISCUSSION THREAD

**Discussion by samvv** [microsoft/pyright#7139] [Q&A] [answered]
I'd like to do the following:

```py
elements: list[str | int] = [ 1, 'foo', 'bar', 3 ]
# Imagine k : int is computed somewhere
elements[k] -= 1
```

I thought this would be easy to solve with one of following expressions:

```py
assert(isinstance(elements[k], int))
elements[k] -= 1
# or
cast(elements[k], int) -= 1
```

But this does not work. How do I solve this issue?

**Comment by erictraut** [comment, marked as answer]
Type narrowing is never applied to index expressions that use variable subscripts (like `elements[k]`) because it's not possible in general for a static type checker to know whether the target of that expression refers to the same element if the expression appears multiple times.

Type narrowing is supported if you use an integer literal value as a subscript.

```python
assert(isinstance(elements[0], int))
elements[0] -= 1
```

For more details about type narrowing and the expression forms that are supported for narrowing, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

Here's a type-safe way to do this using a temporary variable.

```python
elements: list[str | int] = [1, "foo", "bar", 3]
k = 1

x = elements[k]
assert isinstance(x, int)
elements[k] = x - 1
```

**Comment by samvv** [comment]
Oh of course! I could just expand the expression! Why do I keep asking such obvious questions 😅 Thanks!
